File: C++/代码随想录/Binary_Tree/Find the attributes of Binary tree/404. Sum of Left Leaves.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def sumOfLeftLeaves(root):
    if not root:
        return 0
    return dfs(root, False)

def dfs(node, isLeft):
    if not node:
        return 0
    if not node.left and not node.right and isLeft:
        return node.val
    return dfs(node.left, True) + dfs(node.right, False)

# 测试用例
tree1 = TreeNode(3)
tree1.left = TreeNode(9)
tree1.right = TreeNode(20)
tree1.right.left = TreeNode(15)
tree1.right.right = TreeNode(7)

print(sumOfLeftLeaves(tree1))  # 24

tree2 = TreeNode(1)
tree2.left = TreeNode(2)
tree2.right = TreeNode(3)
tree2.left.left = TreeNode(4)
tree2.left.right = TreeNode(5)

print(sumOfLeftLeaves(tree2))  # 4

File: C++/代码随想录/Binary_Tree/Find the attributes of Binary tree/404. Sum of Left Leaves.cpp
// 404. Sum of Left Leaves
// https://leetcode.com/problems/sum-of-left-leaves/
// clang++ 404.\ Sum\ of\ Left\ Leaves.cpp -std=c++17
#include "../../printUtils.hpp"

template <typename T>
T sumOfLeftLeaves(TreeNode<T>* root) {
    if (root == nullptr)
        return T();

    T sum = T();
    if (root->left != nullptr && root->left->left == nullptr && root->left->right == nullptr)
        sum += root->left->val;

    sum += sumOfLeftLeaves(root->left);
    sum += sumOfLeftLeaves(root->right);

    return sum;
}

int main() {
    TreeNode<int>* root1 = buildTree({3, 9, 20, -1, -1, 15, 7}, -1);
    int sum1 = sumOfLeftLeaves(root1);
    std::cout << "Sum of left leaves: " << sum1 << std::endl;
    deleteTree(root1);

    TreeNode<double>* root2 = buildTree({3.0, 9.346, 20.0, -1.0, -1.0, 15.0, 7.0}, -1.0);
    double sum2 = sumOfLeftLeaves(root2);
    std::cout << "Sum of left leaves: " << sum2 << std::endl;
    deleteTree(root2);

    return 0;
}
